README.md
# runstring

[![CI](https://github.com/meyfa/runstring/actions/workflows/main.yml/badge.svg)](https://github.com/meyfa/runstring/actions/workflows/main.yml)
[![Test Coverage](https://api.codeclimate.com/v1/badges/39e38e2764a453e66a43/test_coverage)](https://codeclimate.com/github/meyfa/runstring/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/39e38e2764a453e66a43/maintainability)](https://codeclimate.com/github/meyfa/runstring/maintainability)

Convert JS functions to runnable strings, with parameter serialization!
(We also support TypeScript natively!)

This was made for Electron's `executeJavaScript()` method, so that the code does
not need to be constructed as a string but can be passed as a function.
runstring will convert that function and its parameters to an IIFE string.

The following parameter types are supported:

- Literals `null`, `undefined`, `true`, `false`
- Numbers
- Strings
- Functions (both standard notation and arrow notation)
- Arrays
- Objects

Strings will be escaped. Nesting of values (in objects and arrays) is supported
without limit.

## Usage

### Basic Usage

Simply invoke the module with a function and its parameters to obtain the IIFE
string.

```js
const runstring = require('runstring')

const code = runstring(myFunction, arg1, arg2 /* , ... */)
// do something with `code`
```

### Example 1

```js
const runstring = require('runstring')

const code = runstring(function (a, b) {
  return a + b
}, 5, 7)
```

`code` would now store a string similar to this:
`';(function (a, b) { return a + b })(5, 7);'`. That string could be passed to
Electron's `executeJavaScript()`, or the standard `eval()` (eval is evil, but if
you have your reasons to use it, might as well do it right).

### Example 2

Any parameter type is supported &mdash; numbers, strings, objects, arrays, and
even other functions can all be passed to the module for stringification:

```js
const runstring = require('runstring')

const code = runstring(function (predicate, action) {
  const elements = document.getElementsByClassName('item')
  for (let i = 0; i < elements.length; ++i) {
    if (predicate(elements[i])) {
      action(elements[i])
    }
  }
}, (e) => e.tagName.toLowerCase() === 'div', removeElement)

function removeElement (e) {
  e.parentNode.removeChild(e)
}
```

`code` would now store the following string:

```
';(function (predicate, action) {
  const elements = document.getElementsByClassName('item')
  for (let i = 0; i < elements.length; ++i) {
    if (predicate(elements[i])) {
      action(elements[i])
    }
  }
})((e) => e.tagName.toLowerCase() === 'div', function removeElement (e) {
  e.parentNode.removeChild(e)
});'
```


CHANGELOG.md
# Changelog

## [3.0.2](https://github.com/meyfa/runstring/compare/v3.0.1...v3.0.2) (2023-09-17)


### Bug Fixes

* Use `String.prototype.slice()` instead of deprecated `.substr()` ([#90](https://github.com/meyfa/runstring/issues/90)) ([fc128db](https://github.com/meyfa/runstring/commit/fc128db599a086d5787bec7183f734f25c03f51c))

## [3.0.1](https://github.com/meyfa/runstring/compare/v3.0.0...v3.0.1) (2023-07-23)


### Miscellaneous Chores

* Publish with provenance ([#80](https://github.com/meyfa/runstring/issues/80)) ([dfc8847](https://github.com/meyfa/runstring/commit/dfc8847ff78c628f26bbd8245272c46159da00db))

## [3.0.0](https://github.com/meyfa/runstring/compare/v2.1.0...v3.0.0) (2023-07-05)


### ⚠ BREAKING CHANGES

* Drop support for Node.js older than v18.16.1 ([#76](https://github.com/meyfa/runstring/issues/76))

### Miscellaneous Chores

* Drop support for Node.js older than v18.16.1 ([#76](https://github.com/meyfa/runstring/issues/76)) ([ee6e825](https://github.com/meyfa/runstring/commit/ee6e82586120360eb6777b2797b4e0eacfa4461b))

## [2.1.0](https://github.com/meyfa/runstring/compare/v2.0.1...v2.1.0) (2023-03-23)


### Features

* Improve typings to validate arguments ([#63](https://github.com/meyfa/runstring/issues/63)) ([c84f67c](https://github.com/meyfa/runstring/commit/c84f67c69f5c2dabb175217f3172417621fe0a3c))


test/index.js
/* eslint-disable   no-unused-expressions */
/* global           describe, it */
'use strict'

const index = require('../')

describe('index.js', function () {
  describe('function stringification', function () {
    it('should accept nothing but functions', function () {
      function test (value) {
        try {
          index(value)
        } catch (e) {
          return
        }
        throw new Error('not thrown for type: ' + typeof value)
      }
      test()
      test(null)
      test(true)
      test('function')
      test(42)
      test([1, 2, 3])
      test({ foo: 'bar' })
    })

    it('should return a string', function () {
      const result = index(function () {})
      if (typeof result !== 'string') {
        throw new Error('expected a string but got ' + typeof result)
      }
    })

    it('should start with a semicolon', function () {
      const result = index(function () {})
      if (!/^;/.test(result)) {
        throw new Error('not starting with a semicolon: ' + result)
      }
    })

    it('should end with a semicolon', function () {
      const result = index(function () {})
      if (!/;$/.test(result)) {
        throw new Error('not ending with a semicolon: ' + result)
      }
    })

    it('should be an IIFE', function () {
      const result = index(function () {})
      if (!/\(function .*\)\(.*\)/.test(result)) {
        throw new Error('not an immediately-invoked function expression: ' + result)
      }
    })

    it('should keep the parameters', function () {
      // eslint-disable-next-line no-unused-vars
      const result = index(function (foo, bar) {})
      if (!/function\s*\(foo,\s*bar\)/.test(result)) {
        throw new Error('parameters not kept: ' + result)
      }
    })

    it('should stringify the contents', function () {
      const result = index(function (a, b) {
        return a || b
      })
      if (!/\{\s*return a \|\| b\s*\}/.test(result)) {
        throw new Error('contents not stringified correctly: ' + result)
      }
    })

    it('should stringify arrow functions', function () {
      const result = index((a, b) => {
        return a || b
      })
      if (!/\(a,\s*b\)\s*=>\s*\{\s*return a \|\| b\s*\}/.test(result)) {
        throw new Error('invalid stringification: ' + result)
      }
    })
  })

  describe('parameter serialization', function () {
    it('should serialize undefined', function () {
      const result = index(function (undef0, undef1) {
        return undef0 === undef1
      }, undefined, undefined)
      if (!/\)\(undefined,\s*undefined\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize numbers', function () {
      const result = index(function (a, b) {
        return a + b
      }, 42, 13.37)
      if (!/\)\(42,\s*13\.37\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize strings', function () {
      const result = index(function (a, b) {
        return a + ' ' + b
      }, 'hello', 'world')
      if (!/\)\("hello",\s*"world"\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should escape serialized strings', function () {
      const result = index(function (a, b) {
        return a + ' ' + b
      }, 'foo"bar\\"\n')
      if (!/\)\("foo\\"bar\\\\\\"\\n"\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize booleans', function () {
      const result = index(function (a, b) {
        return a || b
      }, true, false)
      if (!/\)\(true,\s*false\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize null', function () {
      const result = index(function (a) {
        console.log(a)
      }, null)
      if (!/\)\(null\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize objects', function () {
      const result = index(function (a) {
        console.log(a)
      }, {
        foo: 'bar',
        baz: true
      })
      if (!/\)\(\{\s*"?foo"?:\s*"bar",\s*"?baz"?:\s*true\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize arrays', function () {
      const result = index(function (a) {
        console.log(a)
      }, [42, 'foo', true])
      if (!/\)\(\[\s*42,\s*"foo",\s*true\s*\]\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions', function () {
      const result = index(function (aFunc) {
        aFunc('hello world')
      }, function (str) {
        console.log(str)
      })
      if (!/\)\(function\s+\(str\)\s*\{\s*console\.log\(str\)\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions in objects', function () {
      const result = index(function (obj) {
        obj.func('hello world')
      }, {
        func: function (str) {
          console.log(str)
        }
      })
      if (!/\)\(\{\s*"?func"?:\s*function\s+\(str\)\s*\{\s*console\.log\(str\)\s*\},?\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions in arrays', function () {
      const result = index(function (obj) {
        obj.func('hello world')
      }, [
        function (str) {
          console.log(str)
        }
      ])
      if (!/\)\(\[\s*function\s+\(str\)\s*\{\s*console\.log\(str\)\s*\},?\s*\]\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize arrow functions', function () {
      const result = index(function (aFunc) {
        aFunc('hello world')
      }, (str) => {
        console.log(str)
      })
      if (!/\)\(\(str\)\s*=>\s*\{\s*console\.log\(str\)\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should use null for unsupported types', function () {
      const result = index(function (val) {
        return val === null
      }, Symbol('some description'))
      if (!/\)\(null\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })
  })

  describe('evaluation', function () {
    it('should not run on its own', function () {
      let hasRun = false
      index(function () {
        hasRun = true
      })
      if (hasRun) {
        throw new Error('function was run')
      }
    })

    it('should produce evaluatable strings', function () {
      const string = index(function (a, b) {
        return a + b
      }, 5, 7)
      // eslint-disable-next-line no-eval
      const result = eval(string)
      if (result !== 12) {
        throw new Error('wrong eval result: ' + result)
      }
    })
  })
})


test/index.test.js
/* eslint-disable   no-unused-expressions */
/* global           describe, it */

const index = require('../index.js')

describe('index.js', function () {
  describe('function stringification', function () {
    it('should accept nothing but functions', function () {
      function test (value) {
        try {
          index(value)
        } catch (e) {
          return
        }
        throw new Error('not thrown for type: ' + typeof value)
      }
      test(undefined)
      test(null)
      test(true)
      test('function')
      test(42)
      test([1, 2, 3])
      test({ foo: 'bar' })
    })

    it('should return a string', function () {
      const result = index(function () {})
      if (typeof result !== 'string') {
        throw new Error('expected a string but got ' + typeof result)
      }
    })

    it('should start with a semicolon', function () {
      const result = index(function () {})
      if (!/^;/.test(result)) {
        throw new Error('not starting with a semicolon: ' + result)
      }
    })

    it('should end with a semicolon', function () {
      const result = index(function () {})
      if (!/;$/.test(result)) {
        throw new Error('not ending with a semicolon: ' + result)
      }
    })

    it('should be an IIFE', function () {
      const result = index(function () {})
      if (!/\(function .*\)\(.*\)/.test(result)) {
        throw new Error('not an immediately-invoked function expression: ' + result)
      }
    })

    it('should keep the parameters', function () {
      // eslint-disable-next-line no-unused-vars
      const result = index(function (foo, bar) {})
      if (!/function\s*\(foo,\s*bar\)/.test(result)) {
        throw new Error('parameters not kept: ' + result)
      }
    })

    it('should stringify the contents', function () {
      const result = index(function (a, b) {
        return a || b
      })
      if (!/\{\s*return a \|\| b\s*\}/.test(result)) {
        throw new Error('contents not stringified correctly: ' + result)
      }
    })

    it('should stringify arrow functions', function () {
      const result = index((a, b) => {
        return a || b
      })
      if (!/\(a,\s*b\)\s*=>\s*\{\s*return a \|\| b\s*\}/.test(result)) {
        throw new Error('invalid stringification: ' + result)
      }
    })
  })

  describe('parameter serialization', function () {
    it('should serialize undefined', function () {
      const result = index(function (undef0, undef1) {
        return undef0 === undef1
      }, undefined, undefined)
      if (!/\)\(undefined,\s*undefined\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize numbers', function () {
      const result = index(function (a, b) {
        return a + b
      }, 42, 13.37)
      if (!/\)\(42,\s*13\.37\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize strings', function () {
      const result = index(function (a, b) {
        return a + ' ' + b
      }, 'hello', 'world')
      if (!/\)\("hello",\s*"world"\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should escape serialized strings', function () {
      const result = index(function (a, b) {
        return a + ' ' + b
      }, 'foo"bar\\"\n')
      if (!/\)\("foo\\"bar\\\\\\"\\n"\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize booleans', function () {
      const result = index(function (a, b) {
        return a || b
      }, true, false)
      if (!/\)\(true,\s*false\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize null', function () {
      const result = index(function (a) {
        console.log(a)
      }, null)
      if (!/\)\(null\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize objects', function () {
      const result = index(function (a) {
        console.log(a)
      }, {
        foo: 'bar',
        baz: true
      })
      if (!/\)\(\{\s*"?foo"?:\s*"bar",\s*"?baz"?:\s*true\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize arrays', function () {
      const result = index(function (a) {
        console.log(a)
      }, [42, 'foo', true])
      if (!/\)\(\[\s*42,\s*"foo",\s*true\s*\]\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions', function () {
      const result = index(function (aFunc) {
        aFunc('hello world')
      }, function (str) {
        console.log(str)
      })
      if (!/\)\(function\s+\(str\)\s*\{\s*console\.log\(str\)\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions in objects', function () {
      const result = index(function (obj) {
        obj.func('hello world')
      }, {
        func: function (str) {
          console.log(str)
        }
      })
      if (!/\)\(\{\s*"?func"?:\s*function\s+\(str\)\s*\{\s*console\.log\(str\)\s*\},?\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions in arrays', function () {
      const result = index(function (obj) {
        obj.func('hello world')
      }, [
        function (str) {
          console.log(str)
        }
      ])
      if (!/\)\(\[\s*function\s+\(str\)\s*\{\s*console\.log\(str\)\s*\},?\s*\]\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize arrow functions', function () {
      const result = index(function (aFunc) {
        aFunc('hello world')
      }, (str) => {
        console.log(str)
      })
      if (!/\)\(\(str\)\s*=>\s*\{\s*console\.log\(str\)\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should use null for unsupported types', function () {
      const result = index(function (val) {
        return val === null
      }, Symbol('some description'))
      if (!/\)\(null\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })
  })

  describe('evaluation', function () {
    it('should not run on its own', function () {
      let hasRun = false
      index(function () {
        hasRun = true
      })
      if (hasRun) {
        throw new Error('function was run')
      }
    })

    it('should produce evaluatable strings', function () {
      const string = index(function (a, b) {
        return a + b
      }, 5, 7)
      // eslint-disable-next-line no-eval
      const result = eval(string)
      if (result !== 12) {
        throw new Error('wrong eval result: ' + result)
      }
    })
  })
})


test/index.test.ts
import index from '../src/index.js'

describe('index.js', function () {
  describe('function stringification', function () {
    it('should accept nothing but functions', function () {
      function test (value: any): void {
        try {
          index(value)
        } catch (e) {
          return
        }
        throw new Error('not thrown for type: ' + typeof value)
      }
      test(undefined)
      test(null)
      test(true)
      test('function')
      test(42)
      test([1, 2, 3])
      test({ foo: 'bar' })
    })

    it('should return a string', function () {
      const result: string = index(function () {})
      if (typeof result !== 'string') {
        throw new Error('expected a string but got ' + typeof result)
      }
    })

    it('should start with a semicolon', function () {
      const result: string = index(function () {})
      if (!result.startsWith(';')) {
        throw new Error('not starting with a semicolon: ' + result)
      }
    })

    it('should end with a semicolon', function () {
      const result: string = index(function () {})
      if (!result.endsWith(';')) {
        throw new Error('not ending with a semicolon: ' + result)
      }
    })

    it('should be an IIFE', function () {
      const result: string = index(function () {})
      if (!/\(function .*\)\(.*\)/.test(result)) {
        throw new Error('not an immediately-invoked function expression: ' + result)
      }
    })

    it('should keep the parameters', function () {
      const result: string = index(function (foo: any, bar: any) {}, undefined, undefined)
      if (!/function\s*\(foo,\s*bar\)/.test(result)) {
        throw new Error('parameters not kept: ' + result)
      }
    })

    it('should allow optional parameters', function () {
      const result: string = index(function (foo: any, bar?: any) {}, undefined)
      if (!/function\s*\(foo,\s*bar\)/.test(result)) {
        throw new Error('parameters not kept: ' + result)
      }
    })

    it('should allow default parameters', function () {
      const result: string = index(function (foo: any, bar: any = 42) {}, undefined)
      if (!/function\s*\(foo,\s*bar\s*=\s*42\)/.test(result)) {
        throw new Error('parameters not kept: ' + result)
      }
    })

    it('should stringify the contents', function () {
      const result: string = index(function (a: boolean, b: boolean) {
        return a || b
      }, true, false)
      if (!/\{\s*return a \|\| b;?\s*\}/.test(result)) {
        throw new Error('contents not stringified correctly: ' + result)
      }
    })

    it('should stringify arrow functions', function () {
      const result: string = index((a: boolean, b: boolean) => {
        return a || b
      }, true, false)
      if (!/\(a,\s*b\)\s*=>\s*\{\s*return a \|\| b;?\s*\}/.test(result)) {
        throw new Error('invalid stringification: ' + result)
      }
    })
  })

  describe('parameter serialization', function () {
    it('should serialize undefined', function () {
      const result: string = index(function (undef0: any, undef1: any) {
        return undef0 === undef1
      }, undefined, undefined)
      if (!/\)\(undefined,\s*undefined\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize numbers', function () {
      const result: string = index(function (a: number, b: number) {
        return a + b
      }, 42, 13.37)
      if (!/\)\(42,\s*13\.37\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize strings', function () {
      const result: string = index(function (a: string, b: string) {
        return a + ' ' + b
      }, 'hello', 'world')
      if (!/\)\("hello",\s*"world"\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should escape serialized strings', function () {
      const result: string = index(function (a: string, b: string) {
        return a + ' ' + b
      }, 'foo"bar\\"\n-\u0000\u2028\u2029\u0007', 'baz')
      if (!result.includes(')("foo\\"bar\\\\\\"\\n-\\u0000\\u2028\\u2029\\u0007", "baz")')) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize booleans', function () {
      const result: string = index(function (a: boolean, b: boolean) {
        return a || b
      }, true, false)
      if (!/\)\(true,\s*false\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize null', function () {
      const result: string = index(function (a: any) {
        console.log(a)
      }, null)
      if (!result.includes(')(null)')) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize objects', function () {
      const result: string = index(function (a: any) {
        console.log(a)
      }, {
        foo: 'bar',
        baz: true
      })
      if (!/\)\(\{\s*"?foo"?:\s*"bar",\s*"?baz"?:\s*true\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize arrays', function () {
      const result: string = index(function (a: any) {
        console.log(a)
      }, [42, 'foo', true])
      if (!/\)\(\[\s*42,\s*"foo",\s*true\s*\]\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions', function () {
      const result: string = index(function (aFunc: any) {
        aFunc('hello world')
      }, function (str: any) {
        console.log(str)
      })
      if (!/\)\(function\s+\(str\)\s*\{\s*console\.log\(str\);?\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions in objects', function () {
      const result: string = index(function (obj: any) {
        obj.func('hello world')
      }, {
        func: function (str: any) {
          console.log(str)
        }
      })
      if (!/\)\(\{\s*"?func"?:\s*function\s+\(str\)\s*\{\s*console\.log\(str\);?\s*\},?\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize functions in arrays', function () {
      const result: string = index(function (obj: any) {
        obj.func('hello world')
      }, [
        function (str: any) {
          console.log(str)
        }
      ])
      if (!/\)\(\[\s*function\s+\(str\)\s*\{\s*console\.log\(str\);?\s*\},?\s*\]\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should serialize arrow functions', function () {
      const result: string = index(function (aFunc: any) {
        aFunc('hello world')
      }, (str: any) => {
        console.log(str)
      })
      if (!/\)\(\(str\)\s*=>\s*\{\s*console\.log\(str\);?\s*\}\)/.test(result)) {
        throw new Error('incorrect serialization: ' + result)
      }
    })

    it('should use null for unsupported types', function () {
      const result: string = index(function (val: any) {
        return val === null
      }, Symbol('some description'))
      if (!result.includes(')(null)')) {
        throw new Error('incorrect serialization: ' + result)
      }
    })
  })

  describe('evaluation', function () {
    it('should not run on its own', function () {
      let hasRun = false
      index(function () {
        hasRun = true
      })
      if (hasRun) {
        throw new Error('function was run')
      }
    })

    it('should produce evaluatable strings', function () {
      const string: string = index(function (a: number, b: number) {
        return a + b
      }, 5, 7)
      // eslint-disable-next-line no-eval
      const result: number = eval(string)
      if (result !== 12) {
        throw new Error(`wrong eval result: ${result}`)
      }
    })
  })
})


index.js
'use strict'

// HELPER FUNCTIONS

/**
 * Stringify something. This determines the value type and then applies
 * type-specific stringification logic.
 *
 * @param {*} obj The 'thing' that should be stringified.
 * @returns {string} The object, stringified.
 */
function stringify (obj) {
  // see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof
  switch (typeof obj) {
    case 'undefined':
      return 'undefined'
    case 'boolean':
      return obj.toString()
    case 'number':
      return obj.toString()
    case 'string':
      return stringifyString(obj)
    case 'function':
      return obj.toString()
    case 'object':
      return stringifyObject(obj)
    default:
      return 'null'
  }
}

/**
 * Stringify a true object. This is vaguely similar to JSON serialization.
 *
 * @param {object} obj The object to stringify.
 * @returns {string} The object, stringified.
 */
function stringifyObject (obj) {
  if (obj === null) {
    return 'null'
  }
  // array
  if (Object.prototype.toString.call(obj) === '[object Array]') {
    return '[' + Array.prototype.map.call(obj, stringify).join(', ') + ']'
  }
  // plain object
  return '{' + Object.keys(obj).map((key) => {
    return stringify(key) + ': ' + stringify(obj[key])
  }).join(', ') + '}'
}

/**
 * Stringify a string (wrap it in quotes and escape special characters).
 *
 * @param {string} string The string to stringify.
 * @returns {string} The string escaped and wrapped in quotes.
 */
function stringifyString (string) {
  const escMap = {
    '"': '\\"',
    '\\': '\\\\',
    '\b': '\\b',
    '\f': '\\f',
    '\n': '\\n',
    '\r': '\\r',
    '\t': '\\t'
  }
  // eslint-disable-next-line no-control-regex
  return '"' + string.replace(/[\\"\u0000-\u001F\u2028\u2029]/g, (m) => {
    return escMap[m] || '\\u' + (m.charCodeAt(0) + 0x10000).toString(16).substr(1)
  }) + '"'
}

// EXPORT

/**
 * Create a runnable string for the given function.
 *
 * Arguments after the 'func' argument will map one-to-one onto arguments in
 * the resulting string.
 *
 * @param {function} func  The function.
 * @param {*}        args  Function arguments.
 * @returns {string} The runnable string.
 */
function createRunString (func, ...args) {
  if (typeof func !== 'function') {
    throw new Error('expected a function but got ' + typeof func)
  }
  return ';(' + func + ')(' + args.map(stringify).join(', ') + ');'
}

module.exports = createRunString


src/index.ts
// HELPER FUNCTIONS

/**
 * Stringify something. This determines the value type and then applies
 * type-specific stringification logic.
 *
 * @param {*} obj The 'thing' that should be stringified.
 * @returns {string} The object, stringified.
 */
function stringify (obj: any): string {
  // see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof
  switch (typeof obj) {
    case 'undefined':
      return 'undefined'
    case 'boolean':
      return obj.toString()
    case 'number':
      return obj.toString()
    case 'string':
      return stringifyString(obj)
    case 'function':
      return obj.toString()
    case 'object':
      return stringifyObject(obj)
    default:
      return 'null'
  }
}

/**
 * Stringify a true object. This is vaguely similar to JSON serialization.
 *
 * @param {object} obj The object to stringify.
 * @returns {string} The object, stringified.
 */
function stringifyObject (obj: Record<string, any>): string {
  if (obj === null) {
    return 'null'
  }
  // array
  if (Object.prototype.toString.call(obj) === '[object Array]') {
    return '[' + Array.prototype.map.call(obj, stringify).join(', ') + ']'
  }
  // plain object
  return '{' + Object.keys(obj).map((key) => {
    return stringify(key) + ': ' + stringify(obj[key])
  }).join(', ') + '}'
}

/**
 * Stringify a string (wrap it in quotes and escape special characters).
 *
 * @param {string} string The string to stringify.
 * @returns {string} The string escaped and wrapped in quotes.
 */
function stringifyString (string: string): string {
  const escMap: Record<string, string> = {
    '"': '\\"',
    '\\': '\\\\',
    '\b': '\\b',
    '\f': '\\f',
    '\n': '\\n',
    '\r': '\\r',
    '\t': '\\t'
  }
  // eslint-disable-next-line no-control-regex
  return '"' + string.replace(/[\\"\u0000-\u001F\u2028\u2029]/g, (m) => {
    if (escMap[m] !== undefined) return escMap[m]
    return '\\u' + (m.charCodeAt(0) + 0x10000).toString(16).slice(1)
  }) + '"'
}

// EXPORT

/**
 * Create a runnable string for the given function.
 *
 * Arguments after the 'func' argument will map one-to-one onto arguments in
 * the resulting string.
 *
 * @param {Function} func  The function.
 * @param {*}        args  Function arguments.
 * @returns {string} The runnable string.
 */
function createRunString<T extends (...args: any[]) => unknown> (func: T, ...args: Parameters<T>): string {
  if (typeof func !== 'function') {
    throw new Error('expected a function but got ' + typeof func)
  }
  const funcStr: string = Function.prototype.toString.call(func)
  const argsStr: string = args.map(stringify).join(', ')
  return `;(${funcStr})(${argsStr});`
}

export = createRunString
